File: scripts/tests/external_torture_test.py
#!/usr/bin/env python3
"""
External Registry Torture Test Client

This script demonstrates how to use the NSCR external API for comprehensive
registry torture testing. It performs concurrent operations while monitoring
registry state to ensure consistency.

Usage:
    python3 external_torture_test.py --registry-url http://localhost:7000 --username admin --password admin
"""

import argparse
import asyncio
import aiohttp
import json
import time
import random
import logging
import ssl
from typing import Dict, List, Any
from dataclasses import dataclass
from concurrent.futures import ThreadPoolExecutor
import threading

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

@dataclass
class RegistryState:
    """Represents the current state of the registry"""
    timestamp: int
    total_repositories: int
    total_manifests: int
    total_blobs: int
    active_sessions: int
    health_status: str
    repositories: List[Dict[str, Any]]

class NSCRTortureTest:
    """External torture test client for NSCR registry"""
    
    def __init__(self, registry_url: str, username: str = None, password: str = None):
        self.registry_url = registry_url.rstrip('/')
        self.auth = aiohttp.BasicAuth(username, password) if username and password else None
        self.session = None
        self.state_history: List[RegistryState] = []
        self.operation_count = 0
        self.error_count = 0
        self.lock = threading.Lock()
        
        # Log authentication status
        if self.auth:
            logger.info(f"Using authentication: {username}")
        else:
            logger.info("No authentication configured - using anonymous access")
        
    async def __aenter__(self):
        # Configure SSL context to allow self-signed certificates
        ssl_context = ssl.create_default_context()
        ssl_context.check_hostname = False
        ssl_context.verify_mode = ssl.CERT_NONE
        
        connector = aiohttp.TCPConnector(ssl=ssl_context)
        self.session = aiohttp.ClientSession(auth=self.auth, connector=connector)
        return self
        
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self.session:
            await self.session.close()
    
    async def get_registry_state(self) -> RegistryState:
        """Get comprehensive registry state"""
        try:
            async with self.session.get(f"{self.registry_url}/api/registry/state") as response:
                if response.status == 200:
                    data = await response.json()
                    return RegistryState(
                        timestamp=data['timestamp'],
                        total_repositories=data['summary']['totalRepositories'],
                        total_manifests=data['summary']['totalManifests'],
                        total_blobs=data['summary']['totalBlobs'],
                        active_sessions=data['activeSessions']['count'],
                        health_status=data['health']['status'],
                        repositories=data['repositories']
                    )
                else:
                    logger.error(f"Failed to get registry state: {response.status}")
                    return None
        except Exception as e:
            logger.error(f"Error getting registry state: {e}")
            return None
    
    async def get_health(self) -> Dict[str, Any]:
        """Get registry health status"""
        try:
            async with self.session.get(f"{self.registry_url}/api/registry/health") as response:
                if response.status == 200:
                    return await response.json()
                else:
                    logger.error(f"Failed to get health: {response.status}")
                    return None
        except Exception as e:
            logger.error(f"Error getting health: {e}")
            return None
    
    async def get_repository_details(self, repo_name: str) -> Dict[str, Any]:
        """Get detailed repository information"""
        try:
            async with self.session.get(f"{self.registry_url}/api/registry/repositories/{repo_name}") as response:
                if response.status == 200:
                    return await response.json()
                else:
                    logger.error(f"Failed to get repository details for {repo_name}: {response.status}")
                    return None
        except Exception as e:
            logger.error(f"Error getting repository details for {repo_name}: {e}")
            return None
    
    async def get_active_sessions(self) -> Dict[str, Any]:
        """Get active session information"""
        try:
            async with self.session.get(f"{self.registry_url}/api/registry/sessions") as response:
                if response.status == 200:
                    return await response.json()
                else:
                    logger.error(f"Failed to get active sessions: {response.status}")
                    return None
        except Exception as e:
            logger.error(f"Error getting active sessions: {e}")
            return None
    
    def record_operation(self, success: bool):
        """Record operation result"""
        with self.lock:
            self.operation_count += 1
            if not success:
                self.error_count += 1
    
    def get_success_rate(self) -> float:
        """Calculate success rate"""
        with self.lock:
            if self.operation_count == 0:
                return 0.0
            return (self.operation_count - self.error_count) / self.operation_count * 100
    
    async def monitor_registry_state(self, duration: int = 60):
        """Monitor registry state for specified duration"""
        logger.info(f"Starting registry state monitoring for {duration} seconds")
        start_time = time.time()
        
        while time.time() - start_time < duration:
            state = await self.get_registry_state()
            if state:
                self.state_history.append(state)
                logger.info(f"State: repos={state.total_repositories}, "
                          f"manifests={state.total_manifests}, "
                          f"blobs={state.total_blobs}, "
                          f"sessions={state.active_sessions}, "
                          f"health={state.health_status}")
            
            await asyncio.sleep(5)  # Monitor every 5 seconds
    
    async def perform_health_checks(self, duration: int = 60):
        """Perform continuous health checks"""
        logger.info(f"Starting health checks for {duration} seconds")
        start_time = time.time()
        
        while time.time() - start_time < duration:
            health = await self.get_health()
            if health:
                status = health.get('status', 'unknown')
                if status != 'healthy':
                    logger.warning(f"Registry health degraded: {status}")
                    logger.warning(f"Health details: {json.dumps(health, indent=2)}")
                else:
                    logger.debug(f"Registry health: {status}")
            
            await asyncio.sleep(10)  # Check every 10 seconds
    
    async def perform_repository_consistency_checks(self, duration: int = 60):
        """Perform repository consistency checks"""
        logger.info(f"Starting repository consistency checks for {duration} seconds")
        start_time = time.time()
        
        while time.time() - start_time < duration:
            state = await self.get_registry_state()
            if state:
                for repo in state.repositories:
                    repo_name = repo['name']
                    repo_details = await self.get_repository_details(repo_name)
                    
                    if repo_details:
                        # Check for consistency issues
                        tag_count_from_state = repo['tagCount']
                        tag_count_from_details = repo_details['tagCount']
                        
                        if tag_count_from_state != tag_count_from_details:
                            logger.error(f"Inconsistent tag count for {repo_name}: "
                                       f"state={tag_count_from_state}, details={tag_count_from_details}")
                            self.record_operation(False)
                        else:
                            self.record_operation(True)
                        
                        # Check for manifests without digests
                        for tag in repo_details.get('tags', []):
                            if tag.get('hasManifest', False) and not tag.get('digest'):
                                logger.error(f"Manifest without digest for {repo_name}:{tag['tag']}")
                                self.record_operation(False)
                            else:
                                self.record_operation(True)
            
            await asyncio.sleep(15)  # Check every 15 seconds
    
    async def perform_session_monitoring(self, duration: int = 60):
        """Monitor active sessions"""
        logger.info(f"Starting session monitoring for {duration} seconds")
        start_time = time.time()
        
        while time.time() - start_time < duration:
            sessions = await self.get_active_sessions()
            if sessions:
                active_sessions = sessions.get('activeSessions', [])
                total_sessions = sessions.get('totalActiveSessions', 0)
                
                logger.info(f"Active sessions: {total_sessions}")
                
                for session in active_sessions:
                    session_id = session['id']
                    duration_ms = session['duration']
                    blob_count = session['blobCount']
                    
                    logger.debug(f"Session {session_id}: duration={duration_ms}ms, blobs={blob_count}")
                    
                    # Check for long-running sessions
                    if duration_ms > 300000:  # 5 minutes
                        logger.warning(f"Long-running session detected: {session_id} ({duration_ms}ms)")
            
            await asyncio.sleep(20)  # Check every 20 seconds
    
    async def perform_stress_test(self, duration: int = 60, concurrent_requests: int = 10):
        """Perform stress test with concurrent API requests"""
        logger.info(f"Starting stress test for {duration} seconds with {concurrent_requests} concurrent requests")
        start_time = time.time()
        
        async def stress_worker(worker_id: int):
            """Individual stress test worker"""
            while time.time() - start_time < duration:
                try:
                    # Randomly choose an operation
                    operation = random.choice([
                        'get_state',
                        'get_health', 
                        'get_sessions',
                        'get_repository_details'
                    ])
                    
                    if operation == 'get_state':
                        state = await self.get_registry_state()
                        if state:
                            self.record_operation(True)
                        else:
                            self.record_operation(False)
                    
                    elif operation == 'get_health':
                        health = await self.get_health()
                        if health:
                            self.record_operation(True)
                        else:
                            self.record_operation(False)
                    
                    elif operation == 'get_sessions':
                        sessions = await self.get_active_sessions()
                        if sessions:
                            self.record_operation(True)
                        else:
                            self.record_operation(False)
                    
                    elif operation == 'get_repository_details':
                        state = await self.get_registry_state()
                        if state and state.repositories:
                            repo_name = random.choice(state.repositories)['name']
                            details = await self.get_repository_details(repo_name)
                            if details:
                                self.record_operation(True)
                            else:
                                self.record_operation(False)
                        else:
                            self.record_operation(False)
                    
                    # Random delay between operations
                    await asyncio.sleep(random.uniform(0.1, 1.0))
                    
                except Exception as e:
                    logger.error(f"Stress worker {worker_id} error: {e}")
                    self.record_operation(False)
        
        # Start concurrent workers
        workers = [stress_worker(i) for i in range(concurrent_requests)]
        await asyncio.gather(*workers)
    
    def analyze_state_history(self):
        """Analyze state history for anomalies"""
        logger.info("Analyzing state history for anomalies...")
        
        if len(self.state_history) < 2:
            logger.warning("Insufficient state history for analysis")
            return
        
        # Check for sudden changes in metrics
        for i in range(1, len(self.state_history)):
            prev_state = self.state_history[i-1]
            curr_state = self.state_history[i]
            
            # Check for sudden drops in repositories (potential data loss)
            if curr_state.total_repositories < prev_state.total_repositories:
                logger.warning(f"Repository count decreased: {prev_state.total_repositories} -> {curr_state.total_repositories}")
            
            # Check for sudden drops in manifests
            if curr_state.total_manifests < prev_state.total_manifests:
                logger.warning(f"Manifest count decreased: {prev_state.total_manifests} -> {curr_state.total_manifests}")
            
            # Check for health status changes
            if prev_state.health_status == 'healthy' and curr_state.health_status != 'healthy':
                logger.warning(f"Health status degraded: {prev_state.health_status} -> {curr_state.health_status}")
    
    def print_summary(self):
        """Print test summary"""
        logger.info("=== Torture Test Summary ===")
        logger.info(f"Total operations: {self.operation_count}")
        logger.info(f"Errors: {self.error_count}")
        logger.info(f"Success rate: {self.get_success_rate():.2f}%")
        logger.info(f"State snapshots collected: {len(self.state_history)}")
        
        if self.state_history:
            first_state = self.state_history[0]
            last_state = self.state_history[-1]
            logger.info(f"Initial state: repos={first_state.total_repositories}, "
                       f"manifests={first_state.total_manifests}, blobs={first_state.total_blobs}")
            logger.info(f"Final state: repos={last_state.total_repositories}, "
                       f"manifests={last_state.total_manifests}, blobs={last_state.total_blobs}")

async def main():
    parser = argparse.ArgumentParser(description='NSCR External Torture Test')
    parser.add_argument('--registry-url', default='http://localhost:7000',
                       help='Registry URL (default: http://localhost:7000)')
    parser.add_argument('--username', help='Registry username (optional)')
    parser.add_argument('--password', help='Registry password (optional)')
    parser.add_argument('--duration', type=int, default=60,
                       help='Test duration in seconds (default: 60)')
    parser.add_argument('--concurrent-requests', type=int, default=10,
                       help='Number of concurrent requests for stress test (default: 10)')
    parser.add_argument('--test-type', choices=['monitor', 'consistency', 'stress', 'all'],
                       default='all', help='Type of test to run (default: all)')
    
    args = parser.parse_args()
    
    # Authentication is optional - warn if not provided
    if not args.username or not args.password:
        print("Warning: No authentication provided - using anonymous access")
        print("If the registry requires authentication, use --username and --password")
    
    async with NSCRTortureTest(args.registry_url, args.username, args.password) as tester:
        logger.info(f"Starting {args.test_type} torture test for {args.duration} seconds")
        
        if args.test_type in ['monitor', 'all']:
            await tester.monitor_registry_state(args.duration)
        
        if args.test_type in ['consistency', 'all']:
            await tester.perform_repository_consistency_checks(args.duration)
        
        if args.test_type in ['stress', 'all']:
            await tester.perform_stress_test(args.duration, args.concurrent_requests)
        
        # Always run health checks and session monitoring
        await asyncio.gather(
            tester.perform_health_checks(args.duration),
            tester.perform_session_monitoring(args.duration)
        )
        
        # Analyze results
        tester.analyze_state_history()
        tester.print_summary()

if __name__ == '__main__':
    asyncio.run(main())

File: scripts/tests/external_torture_test.rb
#!/usr/bin/env ruby

# External Registry Torture Test Client (Ruby)
# This script demonstrates how to use the NSCR external API for comprehensive
# registry torture testing using Ruby fibers for concurrency.

require 'net/http'
require 'uri'
require 'json'
require 'time'
require 'optparse'
require 'logger'
require 'thread'

# Configure logging
logger = Logger.new(STDOUT)
logger.level = Logger::INFO
logger.formatter = proc do |severity, datetime, progname, msg|
  "#{datetime.strftime('%Y-%m-%d %H:%M:%S')} - #{severity} - #{msg}\n"
end

class NSCRTortureTest
  attr_reader :registry_url, :username, :password, :logger
  attr_accessor :operation_count, :error_count, :state_history

  def initialize(registry_url, username = nil, password = nil)
    @registry_url = registry_url.chomp('/')
    @username = username
    @password = password
    @logger = logger
    @operation_count = 0
    @error_count = 0
    @state_history = []
    @mutex = Mutex.new
    
    # Log authentication status
    if @username && @password && !@username.empty? && !@password.empty?
      logger.info("Using authentication: #{@username}")
    else
      logger.info("No authentication configured - using anonymous access")
    end
  end

  # Make HTTP request with optional authentication
  def make_request(endpoint, method = 'GET')
    uri = URI("#{@registry_url}#{endpoint}")
    http = Net::HTTP.new(uri.host, uri.port)
    
    # Configure SSL/TLS for HTTPS
    if uri.scheme == 'https'
      http.use_ssl = true
      http.verify_mode = OpenSSL::SSL::VERIFY_NONE # Allow self-signed certificates
    end
    
    http.read_timeout = 30
    http.open_timeout = 10

    request = case method.upcase
              when 'GET'
                Net::HTTP::Get.new(uri)
              when 'POST'
                Net::HTTP::Post.new(uri)
              when 'DELETE'
                Net::HTTP::Delete.new(uri)
              else
                Net::HTTP::Get.new(uri)
              end

    # Add authentication only if credentials are provided
    if @username && @password && !@username.empty? && !@password.empty?
      request.basic_auth(@username, @password)
    end
    
    request['Accept'] = 'application/json'
    request['Content-Type'] = 'application/json'

    response = http.request(request)
    
    if response.code.to_i == 200
      JSON.parse(response.body)
    else
      logger.error("HTTP #{response.code}: #{response.message}")
      nil
    end
  rescue => e
    logger.error("Request failed: #{e.message}")
    nil
  end

  # Record operation result
  def record_operation(success)
    @mutex.synchronize do
      @operation_count += 1
      @error_count += 1 unless success
    end
  end

  # Get comprehensive registry state
  def get_registry_state
    data = make_request('/api/registry/state')
    if data
      record_operation(true)
      RegistryState.new(
        timestamp: data['timestamp'],
        total_repositories: data.dig('summary', 'totalRepositories') || 0,
        total_manifests: data.dig('summary', 'totalManifests') || 0,
        total_blobs: data.dig('summary', 'totalBlobs') || 0,
        active_sessions: data.dig('activeSessions', 'count') || 0,
        health_status: data.dig('health', 'status') || 'unknown',
        repositories: data['repositories'] || []
      )
    else
      record_operation(false)
      nil
    end
  end

  # Get registry health status
  def get_health
    data = make_request('/api/registry/health')
    if data
      record_operation(true)
      data
    else
      record_operation(false)
      nil
    end
  end

  # Get repository details
  def get_repository_details(repo_name)
    data = make_request("/api/registry/repositories/#{repo_name}")
    if data
      record_operation(true)
      data
    else
      record_operation(false)
      nil
    end
  end

  # Get active sessions
  def get_active_sessions
    data = make_request('/api/registry/sessions')
    if data
      record_operation(true)
      data
    else
      record_operation(false)
      nil
    end
  end

  # Get blob information
  def get_blob_info
    data = make_request('/api/registry/blobs')
    if data
      record_operation(true)
      data
    else
      record_operation(false)
      nil
    end
  end

  # Calculate success rate
  def success_rate
    return 0.0 if @operation_count == 0
    ((@operation_count - @error_count).to_f / @operation_count * 100).round(2)
  end

  # Monitor registry state using fibers
  def monitor_registry_state(duration)
    logger.info("Starting registry state monitoring for #{duration} seconds")
    start_time = Time.now

    Fiber.new do
      while Time.now - start_time < duration
        state = get_registry_state
        if state
          @state_history << state
          logger.info("State: repos=#{state.total_repositories}, " \
                     "manifests=#{state.total_manifests}, " \
                     "blobs=#{state.total_blobs}, " \
                     "sessions=#{state.active_sessions}, " \
                     "health=#{state.health_status}")
        end

        # Yield control and sleep
        Fiber.yield
        sleep(5)
      end
    end
  end

  # Perform health checks using fibers
  def perform_health_checks(duration)
    logger.info("Starting health checks for #{duration} seconds")
    start_time = Time.now

    Fiber.new do
      while Time.now - start_time < duration
        health = get_health
        if health
          status = health['status']
          if status != 'healthy'
            logger.warn("Registry health degraded: #{status}")
            logger.warn("Health details: #{health.to_json}")
          else
            logger.debug("Registry health: #{status}")
          end
        end

        Fiber.yield
        sleep(10)
      end
    end
  end

  # Perform repository consistency checks using fibers
  def perform_consistency_checks(duration)
    logger.info("Starting repository consistency checks for #{duration} seconds")
    start_time = Time.now

    Fiber.new do
      while Time.now - start_time < duration
        state = get_registry_state
        if state
          state.repositories.each do |repo|
            repo_name = repo['name']
            repo_details = get_repository_details(repo_name)

            if repo_details
              # Check for consistency issues
              tag_count_state = repo['tagCount']
              tag_count_details = repo_details['tagCount']

              if tag_count_state != tag_count_details
                logger.error("Inconsistent tag count for #{repo_name}: " \
                           "state=#{tag_count_state}, details=#{tag_count_details}")
              end

              # Check for manifests without digests
              repo_details['tags']&.each do |tag|
                if tag['hasManifest'] && !tag['digest']
                  logger.error("Manifest without digest for #{repo_name}:#{tag['tag']}")
                end
              end
            end
          end
        end

        Fiber.yield
        sleep(15)
      end
    end
  end

  # Monitor active sessions using fibers
  def monitor_sessions(duration)
    logger.info("Starting session monitoring for #{duration} seconds")
    start_time = Time.now

    Fiber.new do
      while Time.now - start_time < duration
        sessions = get_active_sessions
        if sessions
          active_sessions = sessions['activeSessions'] || []
          total_sessions = sessions['totalActiveSessions'] || 0

          logger.info("Active sessions: #{total_sessions}")

          active_sessions.each do |session|
            session_id = session['id']
            duration_ms = session['duration']
            blob_count = session['blobCount']

            logger.debug("Session #{session_id}: duration=#{duration_ms}ms, blobs=#{blob_count}")

            # Check for long-running sessions
            if duration_ms > 300_000 # 5 minutes
              logger.warn("Long-running session detected: #{session_id} (#{duration_ms}ms)")
            end
          end
        end

        Fiber.yield
        sleep(20)
      end
    end
  end

  # Perform stress test using fibers
  def perform_stress_test(duration, concurrent_requests)
    logger.info("Starting stress test for #{duration} seconds with #{concurrent_requests} concurrent requests")
    start_time = Time.now

    # Create multiple stress worker fibers
    workers = (1..concurrent_requests).map do |worker_id|
      stress_worker(worker_id, start_time, duration)
    end

    workers
  end

  # Individual stress test worker fiber
  def stress_worker(worker_id, start_time, duration)
    Fiber.new do
      while Time.now - start_time < duration
        begin
          # Randomly choose an operation
          operations = %w[get_state get_health get_sessions get_repository_details get_blob_info]
          operation = operations.sample

          case operation
          when 'get_state'
            state = get_registry_state
            logger.debug("Worker #{worker_id}: get_state #{state ? 'success' : 'failed'}")

          when 'get_health'
            health = get_health
            logger.debug("Worker #{worker_id}: get_health #{health ? 'success' : 'failed'}")

          when 'get_sessions'
            sessions = get_active_sessions
            logger.debug("Worker #{worker_id}: get_sessions #{sessions ? 'success' : 'failed'}")

          when 'get_repository_details'
            state = get_registry_state
            if state && !state.repositories.empty?
              repo_name = state.repositories.sample['name']
              details = get_repository_details(repo_name)
              logger.debug("Worker #{worker_id}: get_repository_details #{details ? 'success' : 'failed'}")
            end

          when 'get_blob_info'
            blob_info = get_blob_info
            logger.debug("Worker #{worker_id}: get_blob_info #{blob_info ? 'success' : 'failed'}")
          end

          # Random delay between operations
          Fiber.yield
          sleep(rand(0.1..1.0))

        rescue => e
          logger.error("Stress worker #{worker_id} error: #{e.message}")
          record_operation(false)
        end
      end
    end
  end

  # Analyze state history for anomalies
  def analyze_state_history
    logger.info("Analyzing state history for anomalies...")

    return logger.warn("Insufficient state history for analysis") if @state_history.length < 2

    # Check for sudden changes in metrics
    (1...@state_history.length).each do |i|
      prev_state = @state_history[i - 1]
      curr_state = @state_history[i]

      # Check for sudden drops in repositories (potential data loss)
      if curr_state.total_repositories < prev_state.total_repositories
        logger.warn("Repository count decreased: #{prev_state.total_repositories} -> #{curr_state.total_repositories}")
      end

      # Check for sudden drops in manifests
      if curr_state.total_manifests < prev_state.total_manifests
        logger.warn("Manifest count decreased: #{prev_state.total_manifests} -> #{curr_state.total_manifests}")
      end

      # Check for health status changes
      if prev_state.health_status == 'healthy' && curr_state.health_status != 'healthy'
        logger.warn("Health status degraded: #{prev_state.health_status} -> #{curr_state.health_status}")
      end
    end
  end

  # Print test summary
  def print_summary
    logger.info("=== Torture Test Summary ===")
    logger.info("Total operations: #{@operation_count}")
    logger.info("Errors: #{@error_count}")
    logger.info("Success rate: #{success_rate}%")
    logger.info("State snapshots collected: #{@state_history.length}")

    if @state_history.any?
      first_state = @state_history.first
      last_state = @state_history.last
      logger.info("Initial state: repos=#{first_state.total_repositories}, " \
                 "manifests=#{first_state.total_manifests}, blobs=#{first_state.total_blobs}")
      logger.info("Final state: repos=#{last_state.total_repositories}, " \
                 "manifests=#{last_state.total_manifests}, blobs=#{last_state.total_blobs}")
    end
  end

  # Run all tests using fiber scheduler
  def run_tests(test_type, duration, concurrent_requests)
    logger.info("Starting #{test_type} torture test for #{duration} seconds")

    fibers = []

    case test_type
    when 'monitor'
      fibers << monitor_registry_state(duration)

    when 'consistency'
      fibers << perform_consistency_checks(duration)

    when 'stress'
      fibers.concat(perform_stress_test(duration, concurrent_requests))

    when 'all'
      # Run all test types concurrently
      fibers << monitor_registry_state(duration)
      fibers << perform_health_checks(duration)
      fibers << perform_consistency_checks(duration)
      fibers << monitor_sessions(duration)
      fibers.concat(perform_stress_test(duration, concurrent_requests))
    end

    # Run all fibers using a simple scheduler
    run_fiber_scheduler(fibers, duration)
  end

  # Simple fiber scheduler
  def run_fiber_scheduler(fibers, duration)
    start_time = Time.now
    active_fibers = fibers.dup

    while Time.now - start_time < duration && active_fibers.any?
      active_fibers.each do |fiber|
        begin
          fiber.resume if fiber.alive?
        rescue FiberError
          # Fiber completed or error occurred
        end
      end

      # Remove completed fibers
      active_fibers.reject! { |fiber| !fiber.alive? }

      # Small delay to prevent busy waiting
      sleep(0.01)
    end

    # Ensure all fibers are properly terminated
    active_fibers.each do |fiber|
      fiber.resume while fiber.alive?
    rescue FiberError
      # Fiber completed
    end
  end
end

# Registry state data class
class RegistryState
  attr_reader :timestamp, :total_repositories, :total_manifests, :total_blobs,
              :active_sessions, :health_status, :repositories

  def initialize(timestamp:, total_repositories:, total_manifests:, total_blobs:,
                 active_sessions:, health_status:, repositories:)
    @timestamp = timestamp
    @total_repositories = total_repositories
    @total_manifests = total_manifests
    @total_blobs = total_blobs
    @active_sessions = active_sessions
    @health_status = health_status
    @repositories = repositories
  end
end

# Main execution
def main
  options = {
    registry_url: 'http://localhost:7000',
    username: nil,
    password: nil,
    duration: 60,
    concurrent_requests: 10,
    test_type: 'all',
    log_level: 'INFO'
  }

  OptionParser.new do |opts|
    opts.banner = "Usage: #{$0} [options]"

    opts.on('--registry-url URL', 'Registry URL (default: http://localhost:7000)') do |url|
      options[:registry_url] = url
    end

    opts.on('--username USERNAME', 'Registry username (optional)') do |username|
      options[:username] = username
    end

    opts.on('--password PASSWORD', 'Registry password (optional)') do |password|
      options[:password] = password
    end

    opts.on('--duration SECONDS', Integer, 'Test duration in seconds (default: 60)') do |duration|
      options[:duration] = duration
    end

    opts.on('--concurrent-requests COUNT', Integer, 'Number of concurrent requests for stress test (default: 10)') do |count|
      options[:concurrent_requests] = count
    end

    opts.on('--test-type TYPE', %w[monitor consistency stress all], 'Type of test to run (default: all)') do |type|
      options[:test_type] = type
    end

    opts.on('--log-level LEVEL', %w[DEBUG INFO WARN ERROR], 'Log level (default: INFO)') do |level|
      options[:log_level] = level
    end

    opts.on('-h', '--help', 'Show this help message') do
      puts opts
      exit
    end
  end.parse!

  # Authentication is optional - warn if not provided
  if options[:username].nil? || options[:password].nil?
    puts "Warning: No authentication provided - using anonymous access"
    puts "If the registry requires authentication, use --username and --password"
  end

  # Set log level
  logger.level = case options[:log_level]
                 when 'DEBUG' then Logger::DEBUG
                 when 'INFO' then Logger::INFO
                 when 'WARN' then Logger::WARN
                 when 'ERROR' then Logger::ERROR
                 else Logger::INFO
                 end

  # Create and run torture test
  tester = NSCRTortureTest.new(
    options[:registry_url],
    options[:username],
    options[:password]
  )

  begin
    tester.run_tests(
      options[:test_type],
      options[:duration],
      options[:concurrent_requests]
    )

    # Analyze results
    tester.analyze_state_history
    tester.print_summary

  rescue Interrupt
    logger.info("Test interrupted by user")
    tester.print_summary
  rescue => e
    logger.error("Test failed: #{e.message}")
    logger.error(e.backtrace.join("\n"))
    exit 1
  end
end

# Run main function if script is executed directly
if __FILE__ == $0
  main
end
